### DISCUSSION TITLE
Does setuptools ignore dependencies mentioned in console_scripts and makes them available anyway?

### DISCUSSION THREAD

**Discussion by Luxxii** [pypa/setuptools#2927] [Q&A] [answered]
Hi everyone, i recently encountered a weird behavior when installing a package, which uses `console_scripts`

I already asked it on SO [here](https://stackoverflow.com/questions/70306617/python-setuptools-installs-console-scripts-even-if-optional-dependency-are-not-i)  and found today an [older question](https://stackoverflow.com/questions/67410934/pip-installs-console-scripts-entry-point-even-when-required-extra-is-not-present?noredirect=1&lq=1), describing the same behavior.

Is this an issue in setuptools? Does the `setup.py` needs to be modified specifically for this?  More information and documentation on this would be very beneficial!

Thanks in advance!

**Comment by Luxxii** [comment]
So basically when having a `console_script` entry point with an extra (e.g. `test=testpkg.test:main [numpy]`), no matter whether the extra was specified (or not), it will be installed and made available in console. The script may not be runnable due to the missing dependencies.

Right now this is not a clean way of how it should behave. At least FallenWarrior from SO and myself expected that scripts are not made available in console when not specifying the extra to be installed. 

Maybe this should be made as an issue here or on [pip](https://github.com/pypa/pip)..

**Reply by LewisGaul** [reply]
I have just encountered the same problem, and it seems like a bug or missing functionality, depending on your perspective. A quick search didn't bring up a pip issue, I'm guessing you didn't raise one?

**Reply by Luxxii** [reply]
No i did not raise an issue back then in [pip](https://github.com/pypa/pip). It still is an open issue to me but is not that important in my projects. However, it would be a "nice-to-have-it-working".

**Comment by abravalheri** [comment, marked as answer]
Hi guys, according to the [entry-points specification](https://packaging.python.org/en/latest/specifications/entry-points/) `extra` is deprecated in that context:

> Using extras for an entry point is no longer recommended. Consumers should support parsing them from existing distributions, but may then ignore them. New publishing tools need not support specifying extras. The functionality of handling extras was tied to setuptools’ model of managing ‘egg’ packages, but newer tools such as pip and virtualenv use a different model.

I have the impression that is the reason why `pip` go ahead and creates the script wrapper anyway even if the package was not installed with the extra.

**Reply by Luxxii** [reply]
Hey, i also found this information then, but it does not give us any information on how we should handle such cases. I guess it is not a common use-case and the information is therefor "ignored".

**Reply by abravalheri** [reply]
Yeah, I think that is the rationale behind pip's choice of implementation. 

A quick alternative is to wrap the import statement in a `try...except` block in your scripts and fail with a message error that asks the user to install the package with the right extras.
